plugins/ipfs/docker/dockeripfs.go

```diff
@@ -0,0 +1,547 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gxed/errors"
+	"github.com/ipfs/go-cid"
+	config "github.com/ipfs/go-ipfs/repo/config"
+	serial "github.com/ipfs/go-ipfs/repo/fsrepo/serialize"
+	"github.com/multiformats/go-multiaddr"
+	cnet "github.com/whyrusleeping/go-ctrlnet"
+
+	"github.com/ipfs/iptb/plugins/ipfs"
+	"github.com/ipfs/iptb/testbed/interfaces"
+	"github.com/ipfs/iptb/util"
+)
+
+var ErrTimeout = errors.New("timeout")
+
+var PluginName = "dockeripfs"
+
+var NewNode testbedi.NewNodeFunc
+var GetAttrDesc testbedi.GetAttrDescFunc
+var GetAttrList testbedi.GetAttrListFunc
+
+const (
+	attrIfName = "ifname"
+)
+
+func init() {
+	NewNode = func(dir string, extras map[string]interface{}) (testbedi.TestbedNode, error) {
+		var imagename string
+		var repobuilder string
+
+		if v, ok := extras["image"]; ok {
+			imagename, ok = v.(string)
+
+			if !ok {
+				return nil, fmt.Errorf("Extra `image` should be a string")
+			}
+
+		} else {
+			return nil, fmt.Errorf("No `image` provided")
+		}
+
+		if v, ok := extras["repobuilder"]; ok {
+			repobuilder, ok = v.(string)
+
+			if !ok {
+				return nil, fmt.Errorf("Extra `repobuilder` should be a string")
+			}
+
+		} else {
+			return nil, fmt.Errorf("No `repobuilder` provided")
+		}
+
+		return &Dockeripfs{
+			dir:         dir,
+			image:       imagename,
+			repobuilder: repobuilder,
+		}, nil
+	}
+
+	GetAttrList = func() []string {
+		return append(GetAttrList(), attrIfName)
+	}
+
+	GetAttrDesc = func(attr string) (string, error) {
+		switch attr {
+		case attrIfName:
+			return "docker ifname", nil
+		}
+
+		return ipfs.GetAttrDesc(attr)
+	}
+}
+
+type Dockeripfs struct {
+	image       string
+	id          string
+	dir         string
+	repobuilder string
+	peerid      *cid.Cid
+	apiaddr     multiaddr.Multiaddr
+	swarmaddrs  []multiaddr.Multiaddr
+}
+
+/// TestbedNode Interface
+
+func (l *Dockeripfs) Init(ctx context.Context, agrs ...string) (testbedi.TBOutput, error) {
+	env, err := l.env()
+	if err != nil {
+		return nil, fmt.Errorf("error getting env: %s", err)
+	}
+
+	cmd := exec.Command(l.repobuilder, "init")
+	cmd.Env = env
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", err, string(out))
+	}
+
+	icfg, err := l.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	lcfg, ok := icfg.(*config.Config)
+	if !ok {
+		return nil, fmt.Errorf("Error: GetConfig() is not an ipfs config")
+	}
+
+	lcfg.Bootstrap = nil
+	lcfg.Addresses.Gateway = ""
+	lcfg.Discovery.MDNS.Enabled = false
+
+	err = l.WriteConfig(lcfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, err
+}
+
+func (l *Dockeripfs) Start(ctx context.Context, args ...string) error {
+	if len(args) > 0 {
+		return fmt.Errorf("cannot yet pass daemon args to docker nodes")
+	}
+
+	alive, err := l.isAlive()
+	if err != nil {
+		return err
+	}
+
+	if alive {
+		return fmt.Errorf("node is already running")
+	}
+
+	cmd := exec.Command("docker", "run", "-d", "-v", l.dir+":/data/ipfs", l.image)
```

Comment:
Interesting future work here would be to use the docker api package: https://docs.docker.com/develop/sdk/#sdk-and-api-quickstart